lib/server.go

```diff
@@ -767,6 +791,53 @@ func (srv *Server) _handleGetHeaders(pp *Peer, msg *MsgDeSoGetHeaders) {
 		headers, blockTip.Hash, blockTip.Height, pp)
 }
 
+// GetHeadersForLocatorAndStopHash returns a list of headers given a list of locator block hashes
+// and a stop hash. Note that this may be slow if the block nodes requested are not in the cache.
+func (srv *Server) GetHeadersForLocatorAndStopHash(
+	locator []*BlockHash,
+	stopHash *BlockHash,
+	maxHeadersPerMsg uint32,
+) ([]*MsgDeSoHeader, error) {
+	var headers []*MsgDeSoHeader
+
+	stopNode, stopNodeExists, stopNodeError := srv.blockchain.GetBlockFromBestChainByHashAndOptionalHeight(stopHash, nil, true)
+	// Special case when there is no block locator provided but only a stop hash.
+	if len(locator) == 0 {
+		if stopNodeError != nil || !stopNodeExists || stopNode == nil {
+			return nil, fmt.Errorf("GetHeadersForLocatorAndStopHash: Stop hash provided but no stop node found")
+		}
+		return []*MsgDeSoHeader{stopNode.Header}, nil
+	}
+	startNode, startNodeExists, startNodeError := srv.blockchain.GetBlockFromBestChainByHashAndOptionalHeight(locator[0], nil, true)
+	if startNodeError != nil || !startNodeExists || startNode == nil {
+		return nil, fmt.Errorf("GetHeadersForLocatorAndStopHash: Start hash provided but no start node found")
+	}
+	nextNodeHeight := startNode.Header.Height + 1
+	nextNode, nextNodeExists, nextNodeError := srv.blockchain.GetBlockFromBestChainByHeight(nextNodeHeight, true)
+	if nextNodeError != nil {
+		return nil, fmt.Errorf("GetHeadersForLocatorAndStopHash: Error getting start node by height: %v", startNodeError)
+	}
+	if !nextNodeExists || nextNode == nil {
+		return nil, nil
+	}
+	for ii := uint32(0); ii < maxHeadersPerMsg; ii++ {
```

Comment:
@diamondhands0 - this is the part of the code that slows down header syncing as we individually fetch block nodes by height until we hit the stop hash or have enough headers.

Some alternatives:
1. Instead of using GetBlockFromBestChainByHeight, we could iterate over the height - hash prefix directly. Doing one single badger iteration is probably faster than a bunch of individual ones.
2. We could compare the stop node height to maxHeadersPerMsg + startNode.Height
    - if stop node height  is greater, then we can find the "true" stop node and iterate back backwards. This will be significantly faster since we'll have the block hash.
    - if stop node height is less, then we just iterate backwards from the stop node.

Reply:
I like the second option. Is it difficult to execute? It also isn't a huge issue if header sync is slow and you want to just leave it for now.

Reply:
oh I already implementated #2 - it's much faster.